framework/src/engine/bft/method.ts

```diff
@@ -235,39 +228,48 @@ export class BFTMethod {
 		await votesStore.setWithSchema(EMPTY_KEY, bftVotes, bftVotesSchema);
 	}
 
-	public async getGeneratorKeys(stateStore: StateStore, height: number): Promise<GeneratorKey[]> {
-		const keysStore = stateStore.getStore(MODULE_STORE_PREFIX_BFT, STORE_PREFIX_GENERATOR_KEYS);
-		const { generators: validators } = await getGeneratorKeys(keysStore, height);
+	public async getGeneratorAtTimestamp(
+		stateStore: StateStore,
+		height: number,
+		timestamp: number,
+	): Promise<Validator> {
+		const paramsStore = stateStore.getStore(MODULE_STORE_PREFIX_BFT, STORE_PREFIX_BFT_PARAMETERS);
+		const bftParams = await getBFTParameters(paramsStore, height);
+		const currentSlot = this.getSlotNumber(timestamp);
+		const generator = bftParams.validators[currentSlot % bftParams.validators.length];
+		return generator;
+	}
+
+	public getSlotNumber(timestamp: number): number {
+		const elapsedTime = timestamp - this._genesisTimestamp;
 
-		return validators;
+		return Math.floor(elapsedTime / this._blockTime);
 	}
 
-	public async setGeneratorKeys(stateStore: StateStore, generators: GeneratorKey[]): Promise<void> {
-		const votesStore = stateStore.getStore(MODULE_STORE_PREFIX_BFT, STORE_PREFIX_BFT_VOTES);
-		const bftVotes = await votesStore.getWithSchema<BFTVotes>(EMPTY_KEY, bftVotesSchema);
-		// This assumes bftVotes.blockBFTInfos will contain currently executing block
-		const nextHeight =
-			bftVotes.blockBFTInfos.length > 0
-				? bftVotes.blockBFTInfos[0].height + 1
-				: bftVotes.maxHeightPrevoted + 1;
+	public getSlotTime(slot: number): number {
+		const slotGenesisTimeOffset = slot * this._blockTime;
 
-		const keysStore = stateStore.getStore(MODULE_STORE_PREFIX_BFT, STORE_PREFIX_GENERATOR_KEYS);
-		const nextHeightBytes = utils.intToBuffer(nextHeight, 4);
+		return this._genesisTimestamp + slotGenesisTimeOffset;
+	}
 
-		await keysStore.setWithSchema(nextHeightBytes, { generators }, generatorKeysSchema);
+	public isWithinTimeslot(slot: number, timestamp: number): boolean {
+		return this.getSlotNumber(timestamp) === slot;
 	}
 
-	private _computeValidatorsHash(validators: BFTValidator[], certificateThreshold: bigint) {
-		const activeValidators: ValidatorsHashInfo[] = [];
+	private _computeValidatorsHash(validators: Validator[], certificateThreshold: bigint) {
+		const validatorsHashInfo: ValidatorsHashInfo[] = [];
 		for (const validator of validators) {
-			activeValidators.push({
+			if (validator.bftWeight <= BigInt(0)) {
```

Comment:
I am not sure if this is correct/needed. From the LIP PR `The BFT weight of the validators is allowed to be 0.` and if we already check for negative weights before as per previous comment.